---
title: "Introduction to centraldogma"
output: rmarkdown::pdf_document
vignette: >
  %\VignetteIndexEntry{centraldogma}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
urlcolor: blue
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,  
  comment = "#>"
)
```

```{r setup, include = FALSE}
library(centraldogma)
```

The [centraldogma](https://github.com/rforbiodatascience22/group_10_package)
package replicates the central dogma of molecular biology. More specifically it
replicates the flow of genetic information from DNA to RNA to protein.

The package consists of the following functions:

* `generate_dna` generates a random DNA sequence.
* `transcribe` transcribes a DNA sequence into a RNA sequence.
* `split_codons` splits a RNA sequence into codons.
* `translate` translates a RNA codon sequence into a protein sequence.
* `plot_aa_occurrence` plots the occurrence of each amino acid in a protein
sequence.

Each function is demonstrated below:

### Generate DNA


### Transcribe DNA

Central dogma simulates transcription from DNA to RNA, using ```transcribe()`` functions,   
which takes in a DNA fragment as the argument and replaces Thymine with Uracil. 

* input - string representing DNA
* output - string representing RNA

```{r}
dna_fragment <- "CCATGTTATG"

transcribe(dna_fragment)
```



### Split Codons

Splits a RNA string into codons based on the start reading frame 
Start can be either 1,2 or 3 and the default is 1

```{r}
string <- "ATCGTACGATATGATACAGAGATAGACATATTTAACGG"

split_codons(string, start = 1)
```


### Translate RNA

### Plot amino acid occurences in the protein 


### Example Workflow / Simulation
